Taiko hitsounding enhancements

nekodex opened the pull request:
This PR aims to enhance Taiko hitsounding by adding more variety to in-game sample playback, without diverging too much from the established or being too distracting. This is achieved through a few different methods, listed roughly in increasing order of noticability:
- Subtle pitch randomisation added to hits (to alleviete the 'machine-gun effect' from repeated sample playback)
- Adding (subtle) left/right panning depending on which side the player hits
- Support different hit/rim samples depending on hitobject volume (by default the thresholds are: 0-59 soft, 60-89 medium, 90+ hard, but I envision these being changable by mappers). Similar idea to banks, but with less distinctiveness between each. The goal was to (crudely) simulate how the timbre of real drums change as you hit them harder, allowing for quieter sounding hits during softer sections whilst still being able to belt it out during loud sections.
- Unique samples for strong hits and strong rim hits - I believe 'stable' osu! layered hitfinish on top of strong hits to do something similar, but this PR replaces that layering (for Argon) and handles the strong playback at input time so players can mash it for fun. This also fixes the issue with strong hits being too loud (due to identical sample being played twice) by cancelling playback of the first sample when a double-hit is detected.
- Added a 'flourish' sample to strong rim hits. By default these get added automatically to beatmaps (in reverse, to bias towards the end of combos), leaving at least 2s between each. I envision mappers will be able to add these in manually eventually

Except for pitch randomisation, panning and the double-hit volume fix for strongs, legacy skins should still sound/behave as they did before. New skins made with appropriately named assets would be able to utilize/skin the new features though.

- [x] depends on ppy/osu-resources#255

---

I've created this as a draft PR due to having concerns about whether this is the correct way of going about this or not. I suspect creating and using `TaikoHitSampleInfo` to encapsulate taiko-specific sample playback logic may result in more side effects that I originally anticipated.

There's currently a few issues remaining:
- [ ] Strong hits with non-Argon and user skins don't correctly cancel playback of the first hit correctly
- [x] Override of `x-hitwhistle` and `x-hitfinish` to `taiko-hitwhistle` and `taiko-hitfinish` doesn't work correctly
(at a guess, might be due to `TaikoHitSampleInfo` not being used in places)
- [x] Some failing tests

nekodex committed:
Add support for Argon-specific samples with a fallback hierarchy (ArgonPro -> Argon -> Triangles)

nekodex committed:
Add Taiko hitsound enhancements - unique stong hits, sample layers (based on volume), subtle pitch randomisation and stereo spread for L/R hits

nekodex committed:
Add support for strong rim hit 'flourish' sample

nekodex committed:
Fix sample lookup logic

nekodex committed:
Fix unnecessary type conversion

nekodex committed:
`TaikoSampleInfo` now has logic for 'taiko-*' sample lookups, making this redundant

nekodex committed:
Better variable naming

peppy committed:
Merge branch 'master' into taiko-hitsounding-enhancements

peppy committed:
Fix nullability to match changes to `HitSampleInfo`

peppy committed:
Update resources

peppy commented on osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs:
```diff
@@ -26,16 +27,20 @@ public partial class GameplaySampleTriggerSource : CompositeDrawable
             this.hitObjectContainer = hitObjectContainer;
 
-            InternalChild = hitSounds = new Container<SkinnableSound>
+            InternalChild = AudioMixer = new DrawableAudioMixer
```
This is the first usage of a nested `DrawableAudioMixer` which makes me very uneasy with it existing.

Can you explain why this was added? I can't find anything in PR or commit messages.

nekodex replied:
Nested as in the `Drawable` hierarchy or as in nested mixers? Because I didn't think that was merged yet? This should just be another mixer at the same (top) level as the other (track/sfx) mixers?

It's used in `DrumSampleTriggerSource` to affect panning of the samples without having to set it per `HitObject`.

peppy replied:
Nested as in drawable yeah.

I think if you just want to change panning, you can do this using a `AudioContainer`. Should be a drop-in replacement.

peppy reviewed (commented):
I need some initial answers before doing anything with this PR.

Might be easier to approach this from scratch applying one piece of the puzzle at a time? There's too many changes to lookup logic in here and I'm not sure what is required (and why).

peppy commented:
In addition to the above, I've resolved conflicts and brought this up-to-date, but it doesn't seem like the new samples are playing. Can you double check?

nekodex commented on osu.Game.Rulesets.Taiko/Objects/HitType.cs:
```diff
@@ -18,6 +18,16 @@ public enum HitType
-        Rim
+        Rim,
+
+        /// <summary>
+        /// A <see cref="Hit"/> that can be hit by the centre portion of the drum.
+        /// </summary>
+        StrongCentre,
+
+        /// <summary>
+        /// A <see cref="Hit"/> that can be hit by the rim portion of the drum.
+        /// </summary>
+        StrongRim
```
Hmm, I guess I didn't fully understand the implications of adding these.

I only really used it to affect sample playback in `DrumSamplePlayer`/`DrumSampleTriggerSource`, so can probably be done in a different way in order to avoid adding new `HitType`s.

peppy replied:
I think this is a much-of-a-muchness at this point, until we start to solidify how skinning should work with these new skins. It doesn't change behaviour because there are no internal usages that rely on the initial `Sample?.Get()` lookup path.

peppy committed:
Avoid using `DrawableAudioMixer`

peppy committed:
Merge branch 'master' into taiko-hitsounding-enhancements

peppy committed:
Remove new taiko `HitTypes` and move strong discerning local to usage

peppy committed:
Tidy up `strong` check logic in `DrumSamplePlayer`

peppy committed:
Tidy up balance specification

peppy committed:
Rename trigger source variables to make more sense

Not sure why "centre" was named "hit" instead

peppy committed:
Avoid using silly switch expression thing we don't use

peppy committed:
Remove `LookupNames` from `GetHashCode` implementation

This is a computed value and should not be considered as part of the hash

peppy committed:
Ensure `CreateHitSampleInfo` creates the correct derived type for taiko

I'm not confident in the direction of using the derived `HitSampleInfo`
type yet, but this will fix the editor creating incorrect samples.

peppy committed:
Convert all taiko hit sound info to use `TaikoHitSampleInfo` for now

peppy committed:
Fix one more case of incorrect `HitSampleInfo` type

peppy committed:
Convert fallback logic to not use switch expression

peppy committed:
Add an explicit `VelocitySuffix` to correctly differentiate between explicit samples

peppy committed:
Refactor sample addition code for efficiency and legibility

peppy commented:
I think I have this PR in a state where I'd be interested in other people reviewing it. Obviously I've refactored too much to make a call on it myself. @bdach you can add this to your bucket list (lower priority; goal to get it in the next-next-release in 2 weeks or so).

peppy committed:
Update resources

bdach commented on osu.Game.Rulesets.Taiko/Audio/TaikoHitSampleInfo.cs:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+
+        public override IEnumerable<string> LookupNames
```
I will say that the amount of complexity in this here property does scare me a little bit. It seems to be working okay, but the logic seems rather brittle and prone to breakage, and I'm not even sure if it can be tested properly. Would be nice if it were tested, though.

bdach commented:
Test failures are all caused by the addition of the "flourish" sample.

`TestSceneSampleOutput`:

![1685778345](https://github.com/ppy/osu/assets/20418176/e8f383b0-2d4a-4790-9898-6a70db5caa4d)

`LegacyBeatmapEncoderTest.TestEncodeDecodeStabilityDoubleConvert("Resources/sample-beatmap-taiko.osu")`:

```diff
diff --git a/1.json b/2.json
index 7b9b3e5..c7a6bfa 100644
--- a/1.json
+++ b/2.json
@@ -726,6 +726,12 @@
         "bank": "normal",
         "volume": 100
       },
+      {
+        "velocity_suffix": "",
+        "name": "taiko-strong-flourish",
+        "bank": "normal",
+        "volume": 100
+      },
       {
         "velocity_suffix": "",
         "name": "taiko-strong-flourish",
@@ -762,6 +768,12 @@
         "bank": "normal",
         "volume": 100
       },
+      {
+        "velocity_suffix": "",
+        "name": "taiko-strong-flourish",
+        "bank": "normal",
+        "volume": 100
+      },
       {
         "velocity_suffix": "",
         "name": "taiko-strong-flourish",
```

The first one will be rather easy to fix, but the second will be annoying.

peppy committed:
Share existing constant for strong hit window

peppy commented on osu.Game/Skinning/ArgonProSkin.cs:
```diff
@@ -24,9 +24,11 @@ public class ArgonProSkin : ArgonSkin
             foreach (string lookup in sampleInfo.LookupNames)
             {
-                string remappedLookup = lookup.Replace(@"Gameplay/", @"Gameplay/ArgonPro/");
+                var sample = Samples?.Get(lookup)
+                             ?? Resources.AudioManager?.Samples.Get(lookup.Replace(@"Gameplay/", @"Gameplay/ArgonPro/"))
+                             ?? Resources.AudioManager?.Samples.Get(lookup.Replace(@"Gameplay/", @"Gameplay/Argon/"))
+                             ?? Resources.AudioManager?.Samples.Get(lookup);
```
I meant to double-check on this. I'm not sure the `Resources.AudioManager` calls should be required.

peppy replied:
I think this is a much-of-a-muchness at this point, until we start to solidify how skinning should work with these new skins. It doesn't change behaviour because there are no internal usages that rely on the initial Sample?.Get() lookup path.

peppy committed:
Update `InvalidOperationException` with new method signature

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy commented on osu.Game.Rulesets.Taiko/Objects/TaikoHitObject.cs:
```diff
@@ -21,5 +24,13 @@ public abstract class TaikoHitObject : HitObject
+
+        public override HitSampleInfo CreateHitSampleInfo(string sampleName = HitSampleInfo.HIT_NORMAL)
+        {
+            if (Samples.FirstOrDefault(s => s.Name == HitSampleInfo.HIT_NORMAL) is HitSampleInfo existingSample)
+                return existingSample.With(newName: sampleName);
```
I copied this across from the `base` implementation. But I agree that it should be updated to return a `TaikoHitSampleInfo`

peppy committed:
Fix `TaikoHitObject`'s `CreateHitSampleInfo` override to always return a `TaikoHitSampleInfo`

peppy committed:
Rename `VelocitySuffix` to `VolumeSuffix`

peppy committed:
Move `FlushPlayback` to base class (and rename to more appropriate method name)

bdach committed:
Rename `get{Velocity -> Volume}Suffix`

In line with the rename of the corresponding property.

peppy committed:
Fix failing tests due to flourish presence

bdach commented:
I would say that this looks ready to go, were it not for the last remaining checkbox from the OP:

> Strong hits with non-Argon and user skins don't correctly cancel playback of the first hit correctly

Is this still relevant? In my testing this branch seems to do no worse than `master`, so maybe it isn't? Please advise.

nekodex commented:
> I would say that this looks ready to go, were it not for the last remaining checkbox from the OP:
> 
> > Strong hits with non-Argon and user skins don't correctly cancel playback of the first hit correctly
> 
> Is this still relevant? In my testing this branch seems to do no worse than `master`, so maybe it isn't? Please advise.

Investigating this again, from what I can tell this is partially resolved:
- For player input, on a 'double hit', it cancels playback of the first hit as expected.
- But with the Autoplay mod on, it doesn't cancel playback and just plays both samples simultaneously.

I'm not sure how Autoplay works, but if it bypasses `DrumSamplePlayer`, then that would likely explain it bypassing the sample playback cancellation logic?

With pitch randomisation being added, the increased volume from waveform stacking is somewhat mitigated, although in some cases it can cause phasing issues instead (which has the effect of dropping perceived volume). In terms of fixing the volume spike, it's 'better' than `master`, but when phasing occurs you get a volume dip instead (and well, it just sounds bad), so that's not ideal. I'll leave you guys to determine whether this is worth fixing now or leaving for later.

However, while investigating this I noticed a behavioural change with how skin hitsound lookups are handled -- on `master`, if a skin doesn't have taiko-specific hitsounds, it will fallback to the stable/classic samples for taiko. With this branch, it instead falls back to the skin's standard-mode equivalents.

peppy commented:
> I'm not sure how Autoplay works, but if it bypasses `DrumSamplePlayer`, then that would likely explain it bypassing the sample playback cancellation logic?

I don't believe it does, but it's probably also hitting both hits on the same frame which may lead to a timing issue with the "Stop playing" operation. I'll look into it today.

> However, while investigating this I noticed a behavioural change with how skin hitsound lookups are handled -- on `master`, if a skin doesn't have taiko-specific hitsounds, it will fallback to the stable/classic samples for taiko. With this branch, it instead falls back to the skin's standard-mode equivalents.

I'll check this one out.

peppy committed:
Move override to avoid touching `PlaySamples`

peppy committed:
Schedule `StopAllPlayback` to match `PlaySamples`' schedule

peppy commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/TaikoLegacySkinTransformer.cs:
```diff
@@ -159,31 +156,5 @@ private string getHitName(TaikoSkinComponents component)
-
-        public override ISample? GetSample(ISampleInfo sampleInfo)
-        {
-            if (sampleInfo is HitSampleInfo hitSampleInfo)
-                return base.GetSample(new LegacyTaikoSampleInfo(hitSampleInfo));
-
-            return base.GetSample(sampleInfo);
```
The reasoning for the change in lookup precedence is the removal of this.

Because this code is in the transformer, each call to `base.GetSample` was doing a full hierarchical lookup. First with `taiko-` prefixes and *then* falling back to non-`taiko-` samples.

With the new code, this has changed to do lookups for `taiko-` and then non-`taiko-` on a *per skin* basis, which matches watch @nekodex reported seeing changed.

I haven't yet looked into a resolution for this, but I'd guess we need to bring this code back here, and apply the `taiko-` prefix locally rather than in `TaikoHitSampleInfo`.

peppy replied:
For sure. I think this is tested in osu! ruleset, but not taiko.

peppy replied:
I've started to look into a fix for this, but it looks like it will be a complete revert of `TaikoHitSampleInfo`. It doesn't need to exist and is only adding complexity.

Before I address this though, I'm waiting for feedback from @nekodex regarding sample banks.

In stable skins, taiko has normal/soft/drum banks, but in this new implementation only one bank exists. But then there's a new "velocity" suffix which is being determined from sample control point volume. This all feels like an oversight to me, so I want to know what the path forward is here before making changes (as they may change the structure of the sample filenames).

peppy committed:
Merge branch 'master' into taiko-hitsounding-enhancements

peppy committed:
As an initial sanity step, fix lookups using `TaikoHitSampleInfo`

peppy committed:
Revert "As an initial sanity step, fix lookups using `TaikoHitSampleInfo`"

This reverts commit 8c4d2d55612131bb9f8899738eebcc2750506821.

peppy committed:
Revert logic in `TaikoLegacySkinTransformer`

peppy committed:
Nuke `TaikoHitSampleInfo` and return to a better known state

peppy committed:
Update one remaining case of `new HitSampleInfo`

peppy committed:
Remove pitch fluctuation (sounds bad)

peppy committed:
Fix `LegacyHitSampleInfo` throwing away `suffix`

bdach commented on osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:
```diff
@@ -536,7 +543,7 @@ public LegacyHitSampleInfo(string name, string? bank = null, int volume = 0, int
 
             public sealed override HitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<string?> newSuffix = default, Optional<int> newVolume = default)
-                => With(newName, newBank, newVolume);
+                => new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered);
```
This change is causing test failures and rightly so because there are several issues with doing this.

This overload is heavily used by several other places in the code (not entirely sure why, looks like something to do with the overload resolution algorithm, probably because it has less default arguments or something). Why is this a problem? Say that you have a `LegacyHitSampleInfo` with `CustomSampleBank = 2`, for instance. Calling this `With()` overload chooses the ctor with `suffix` in it and _not_ the one with `customSampleBank` in it. And the ctor with `suffix` in it, contrary to the one with `customSampleBank` in it, does not set the new `LegacyHitSampleInfo`'s `CustomSampleBank`. This is depended on in a few places (`LegacyBeatmapSkin` first and foremost) and breaks behaviour.

This could be _possibly_ fixable by something along the lines of

```diff
diff --git a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
index ec9efcd7e3..e4fd1b106d 100644
--- a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
+++ b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
@@ -543,7 +543,11 @@ public LegacyHitSampleInfo(string name, string? bank = null, int volume = 0, int
             }
 
             public sealed override HitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<string?> newSuffix = default, Optional<int> newVolume = default)
-                => new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered);
+            {
+                return newSuffix.HasValue
+                    ? new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), newSuffix.GetOr(Suffix), IsLayered)
+                    : new LegacyHitSampleInfo(newName.GetOr(Name), newBank.GetOr(Bank), newVolume.GetOr(Volume), CustomSampleBank, IsLayered);
+            }
 
             public virtual LegacyHitSampleInfo With(Optional<string> newName = default, Optional<string> newBank = default, Optional<int> newVolume = default,
                                                     Optional<int> newCustomSampleBank = default,

```

except with more cases, but problem number 2 is `FileHitSampleInfo` which derives from `LegacyHitSampleInfo`. The way this overriding and double-pathery works is fundamentally incompatible with inheritance and I don't know how you fix it without a rewrite.

With all of that in mind - what was the point of the `Suffix` thing and the entire `CustomSampleBank` / `Suffix` dual ctor path thing? What was it trying to achieve? Because if I'm honest I find it weird and confusing and would much rather not have it exist.

Probably not much I can review further here until we get that straightened out.

peppy replied:
I'm not really sure it needs to exist, and I hope it doesn't because I share the same feelings. The inheritance is a mess.

Maybe it's to make saving work? Will require further investigation.

peppy replied:
I misread your question, so here's a proper answer:

> With all of that in mind - what was the point of the Suffix thing and the entire CustomSampleBank / Suffix dual ctor path thing? What was it trying to achieve? Because if I'm honest I find it weird and confusing and would much rather not have it exist.

Basically, without it the suffix gets ignored. See the original code which accepts a suffix in, but does not pass it through.

But given the feedback from @nekodex, this may have been a red herring as the suffix I was applying is much later in the process, so I think it's probably fine to revert this change.

peppy replied:
Yeah, I've confirmed this change isn't required, so will revert immediately.

nekodex commented:
From Discord discussion:

The volume-suffixes don't appear to be working correctly on some beatmaps, e.g. `Inner Oni` of https://dev.ppy.sh/beatmapsets/921535#taiko/1924484

`taiko-hitnormal-soft` is being used instead of `taiko-hitnormal-hard`, so sections are mixing soft `hitnormal` samples with hard `hitclap` ones, which sounds weird.

peppy commented:
This is because of stupid shit. Basically, the preloading logic for samples happens using the `HitObject.Samples`, which means it doesn't catch anything done by a skin transformer (since that's at a `DrawableHitObject` level at earliest).

I'll return to the drawing board.

peppy committed:
Revert "Fix `LegacyHitSampleInfo` throwing away `suffix`"

This reverts commit 2aadad2242df5e65ebcec0bb8f374daacbe917a4.

peppy commented:
So the best way forward I can propose here is that we don't add these new `soft` / `normal` / `loud` versions and stick to what works. `normal` `soft` and `drum`. Rename the sounds and apply the conversion at the taiko transformer to get things in an initial sane state.

peppy committed:
Add ignore rules for finish/whistle in code rather than resources

peppy committed:
Move volume suffixes to banks instead

peppy committed:
Refactor strong handling to only apply to argon and to preload correctly

peppy committed:
Update lookup logic

peppy committed:
Nuke all the volume weird logic

peppy committed:
Add bank for flourishes

peppy committed:
Try and make strong hits work again

peppy committed:
Fix beatmap encoder stuff

bdach commented:
I looked into the [latest breakage](https://discord.com/channels/188630481301012481/188630652340404224/1121770583181496440) mentioned on discord, and I think I might have an explanation, but first, just to confirm:

@peppy is the breakage you were seeing that some strong hits are playing the wrong samples (i.e. `normal-hitfinish` instead of `taiko-strong-hitnormal`)? If yes, were you hitting the strong hits with both keys (or using autoplay), and if yes, can you verify that hitting the strong hits with a single key does not trigger this?

If all of the above holds (and it does for me locally), then the cause of this breakage is the intersection of `GameplaySampleTriggerSource.GetMostValidObject()` logic, and some `DrumSamplePlayer` logic added here. The explanation goes something like this:

When a strong hit is hit with both keys, it triggers `GameplaySampleTriggerSource` two times. The first time, the strong hit is entirely unjudged, and so `GetMostValidObject()` returns the strong hit, samples are taken from therein, and everything is all fine and dandy. When the _second_ hit happens (even if it is the same frame), the strong hit is now judged (because it was hit, its _strong nested object_ isn't, at most), so `GetMostValidObject()` actually returns _the next object from the beatmap now_. Which may not be necessarily strong at all.

What adds insult to injury here is some new logic added in this PR:

https://github.com/ppy/osu/blob/631e2927739eeb1e8285bd6b75ec1b641b70de4f/osu.Game.Rulesets.Taiko/UI/DrumSamplePlayer.cs#L45
https://github.com/ppy/osu/blob/631e2927739eeb1e8285bd6b75ec1b641b70de4f/osu.Game.Rulesets.Taiko/UI/DrumSamplePlayer.cs#L73-L77

This is trying to manually reimplement the 30ms strong hit window. But do note that the way this is written is _entirely time-based_. There is no notion of `GetMostValidHitObject()` here, or _any_ object. So what happens is:

- First hit of the strong object plays correct sample
- The second hit plays a wrong sample, because `GetMostValidHitObject()` change...
- ...and because the logic mentioned above is time-based and is not aware that `GetMostValidHitObject()` changed, it _flushes the playback of the first strong (correct) sample_, thinking that it's the same sample and shouldn't layer.

It's late so I'm leaving it there, as I have no ideas how to fix this mess at present. Hopefully this investigation helps somewhat.

peppy commented:
Replaced by https://github.com/ppy/osu/pull/24119.
